ops/model.py

```diff
@@ -810,6 +901,291 @@ def __init__(self, name: str, address_info: '_AddressDict'):
+        self.expires = expires
+        self.rotation = rotation
+        self.rotates = rotates
+
+    @classmethod
+    def from_dict(cls, id: str, d: '_SerializedData') -> 'SecretInfo':
+        """Create new SecretInfo object from ID and dict parsed from JSON."""
+        expires = typing.cast(Optional[str], d.get('expires'))
+        try:
+            rotation = SecretRotate(typing.cast(Optional[str], d.get('rotation')))
+        except ValueError:
+            rotation = None
+        rotates = typing.cast(Optional[str], d.get('rotates'))
+        return cls(
+            id=id,
+            label=typing.cast(Optional[str], d.get('label')),
+            revision=typing.cast(int, d['revision']),
+            expires=timeconv.parse_go_timestamp(expires) if expires is not None else None,
+            rotation=rotation,
+            rotates=timeconv.parse_go_timestamp(rotates) if rotates is not None else None,
+        )
+
+    def __repr__(self):
+        return ('SecretInfo('
+                'id={self.id!r}, '
+                'label={self.label!r}, '
+                'revision={self.revision}, '
+                'expires={self.expires!r}, '
+                'rotation={self.rotation}, '
+                'rotates={self.rotates!r})'
+                ).format(self=self)
+
+
+class Secret:
+    """Represents a single secret in the model.
+
+    This class should not be instantiated directly, instead use
+    :meth:`Model.get_secret` (for consumers and owners), or
+    :meth:`Application.add_secret` or :meth:`Unit.add_secret` (for owners).
+
+    All secret events have a :code:`.secret` attribute which provides the
+    :class:`Secret` associated with that event.
+    """
+
+    _key_re = re.compile(r'^([a-z](?:-?[a-z0-9]){2,})$')  # copied from Juju code
+
+    def __init__(self, backend: '_ModelBackend',
+                 id: Optional[str] = None,
+                 label: Optional[str] = None,
+                 content: Optional[Dict[str, str]] = None):
+        if not (id or label):
+            raise TypeError('Must provide an id or label, or both')
+        if id is not None:
+            id = self._canonicalize_id(id)
+        self._backend = backend
+        self._id = id
+        self._label = label
+        self._content = content
+
+    def __repr__(self):
+        fields = []  # type: List[str]
+        if self._id is not None:
+            fields.append('id={!r}'.format(self._id))
+        if self._label is not None:
+            fields.append('label={!r}'.format(self._label))
+        return '<Secret ' + ' '.join(fields) + '>'
+
+    @classmethod
+    def _canonicalize_id(cls, id: str) -> str:
+        """Return the canonical form of the given secret ID, with the 'secret:' prefix."""
+        id = id.strip()
+        if not id.startswith('secret:'):
+            id = 'secret:' + id  # add the prefix if not there already
+        return id
+
+    @classmethod
+    def _validate_content(cls, content: Optional[Dict[str, str]]):
```

Comment:
I kinda disagree here that this is an issue. For a start, if `_validate_content` (or whatever we call it) returns a bool, you can't give good error messages when you do raise.

I don't think we need a way for the user to ask is this a valid secret. You can just add it and the OF will check for you and you'll get an appropriate `TypeError` or `ValueError` if it's no good. That said, we can always add a user-facing function later if there's a compelling use case.

I don't think it needs to be invoked from the constructor, as "This class should not be instantiated directly" as the docstring says. The `Secret` class is difficult to instantiate directly anyway, as the first argument is `backend`, which the user doesn't really have access to (backend implementations and accessors are underscore-private). So the user won't be calling `Secret()` directly, and every place we call it with `content` we validate the content first. We could do it in the constructor *as well*, but that would duplicate effort.

Reply:
I think this is putting way too much faith into users reading the docs and/or using an editor with LSP hinting.

Even if we don't make a public method which allows users to validate "is this a valid `dict` to turn into a secret" which returns a bool (and logs messages along with returning `False` if it's not), this sort of validation should absolutely be invoked from the constructor.

It is not hard to instantiate directly. Python doesn't really enforce this, the constructor doesn't check, and it's trivial for users to just pass `None`, even if they only do it once, to say nothing of the fact that even `__foo` is still accessible from people writing code. As a user of OF, I don't want yet another place in my code where I need to potentially catch exceptions which OF should be handling (and re-raising as a more specific exception, like `ModelError` or whatever), and I definitely do not want to do it for a case like this.

Frankly, the only reason it's duplicating effort to put it into the constructor is because OF itself explicitly calls it before `add_secret`, and in `set_content`, this gets called, but there also, it checks the dict against an arbitrary, private, class-level regexp `_key_re = re.compile(r'^([a-z](?:-?[a-z0-9]){2,})$')`, and if it fails, OF throws a `ValueError` or `TypeError`. It's very, very possible for a relatively new user of Secrets to run into multiple exceptions (trying to pass a non-`str` as a value, trying a disallowed key format, etc) before actually encountering a usable value.

It's just a bad user experience. It is extremely safe, and easy, to both move the validation into the constructor itself instead of calling it from OF code before constructing an object (and logically identical). It's also easy to throw a `InvalidSecretError`  with a message which includes the complete allowed format before users hit one or the other, and "the docs are clear" and/or "the Juju docs specify the format" probably aren't discoverable enough when it's this simple to make it better.

Reply:
Hmm ... I don't mind improving the experience here, but two things:

1) By itself this doesn't make it good, but there's a precedent for this, for example `Relation` and `RelationData` and other model classes say the same thing, "This class should not be instantiated directly", but don't prevent you from instantiating it with a dummy backend and/or invalid data attributes if you really want (Python's "we're all consenting adults" philosophy).
2) In practice, it seems to me that if you were blindly trying to call `Secret()` to instantiate a secret object, you'd at least have to get the args right, and your editor would show you that "backend" is a required argument. So you'd be guided to think "What do I need to supply for 'backend'? I'll read the docstring ... oh, I see, I should be calling add_secret or similar."

Has the issue you're describing been (or will this be) a problem in practice? Have we run into times where people are actually trying to instantiate (say) a `Relation` directly, with an explicit `None` backend, and things going wrong?

If we consider improving this we should at least know whether it's a real problem today, and then consider it on the larger scale of also improving the situation for the existing classes like `Relation` (in a separate PR).

Reply:
Oh, one other thing: it's actually logically different to call it in the constructor, because currently for example in `add_secret` we want to call it before calling the hook tool, whereas the `Secret(...)` call is only called after that. So if we added it to the constructor you'd still need it in both places.

Reply:
1. I mean, yes, it's Python, so you can do a lot of things. Just because it may be one way doesn't mean that it still can't be as safeguarded as possible, though. It only appears to be used in a few places. 
  - `[Application|Unit].add_secret` (in both cases moving to the constructor is an easy step and logically identical). 
  - `Secret.set_content` (in which case the validation could raise a more explicit exception with a clear message). It's already past the constructor here. 
  - My suggestion that "we could call it from the constructor" doesn't necessarily mean "move the code to the constructor". Just to call the method from inside the constructor explicitly rather than before it in `[Application|Unit].add_secret`, which is more or less -2 lines (in each `add_secret`) and +1 (in `__init__`). It's pretty small and adds a lot of "training wheels"
  2. In practice, yes, sort-of. The assumption that users actually read and/or pay attention to what their editor suggests in dynamically-typed languages is generous, in my experience, but even so, it would be easy for OF users to run face-first into this when trying to `set_content` if they violated an invisible regexp (where the docstring is not nearly as clear as the messages in the exceptions raised by `_validate_content`) in any case.
  
Yes, we have run into people trying to instantiate, say, a `Relation` directly, albeit mostly people who are new to charming and/or OF. 

More directly, we (and I) have been impacted by needing to add an annoyingly high amount of exception guarding in charm code from behavior changes in OF (in particular, `relation_get` suddenly being able to return `None`). Open source is a democracy/consensus model, and sometimes pain points are deemed to be too much, and sometimes they're acceptable consequences. I mean, you don't necessarily have to implement the suggestions here. I'm not always right, but I always have an opinion ;) 

My opinion in this case is that the amount of work needed to "bubble wrap" things a little bit and provide more meaningful exceptions is small in terms of the potential time saved for charm authors, but "considered, and I'm not going to, they can read the docs" is ok, too. It's your call, ultimately.

In practice, `_key_re` requires `_validate_content` to be a `@classmethod` to access it, but it's never accessed by anything else, and could just as easily be a top-level constant if we really wanted to compile it only once, or a method-level variable anyway, in which case `_validate_content` could also be `@staticmethod` anyway.

The desire to call it prior to the hook is true, but the constructor for `Secret` has `id` as optional anyway. Again, in practice, there's not a practical difference between instantiating `Secret` first and setting `Secret._id` after the hook tool versus the converse except for your preference, and that doing the inverse adds input validation to the constructor. Your choice, in the end.

Reply:
Regarding "logically identical", maybe I'm missing what you're saying here, but moving the `_validate_content` step to the constructor wouldn't work for the two `add_secret` methods, because you need to validate before calling the hook tool. I could create the `Secret` object at the top of those functions, but like you suggest, you'd have to poke the `id` in afterwards, so it's awkward. And for `get_secret` it would work, but mean we're unnecessary validating content we just received from Juju.

> More directly, we (and I) have been impacted by needing to add an annoyingly high amount of exception guarding in charm code from behavior changes in OF (in particular, relation_get suddenly being able to return None).

This is a fair point in general. I'm hoping to be very careful with backwards compatibility as I shepherd `ops`. Even for the upcoming 2.0 release, which is a major version bump and can be backwards-incompatible, we're planning to be extremely careful about changing stuff, and almost all charms should continue to work as is. The main "breakage" is not API at all, but bumping the minimum Python version up to 3.8.

> My opinion in this case is that the amount of work needed to "bubble wrap" things a little bit and provide more meaningful exceptions is small in terms of the potential time saved for charm authors, but "considered, and I'm not going to, they can read the docs" is ok, too. It's your call, ultimately.

It's a fair point of view. My take is that it's a bit of a false sense of security, and if people are just blindly providing dummy values for things like `backend` (not to mention ignoring docs), they need to up their coding game. :-)

Another counterpoint is that you can only go so far with more and more checks before the implementation (and even sometimes the usage) is unwieldy. In future we're planning to lean in harder on static type checking. But in the meantime we're striving for a happy medium of a few runtime checks as appropriate.

All that said, I've considered your point of view, but decided not to make this specific change now.